### ISSUE TITLE
Sticky header in a useReactTable table with useVirtualizer disappears when scrolling

### ISSUE THREAD

**Issue by wjthieme** [TanStack/virtual#640]
### Describe the bug

When using a react-table together with vritualizer and a sticky header the sticky header disappears when scrolling. This is due to the fact that when combining the two the table should be wrapped in two divs:
1. The container that takes the height that should be taken up. This is the div that is scrollable
2. The div that directly wraps the container and takes the height of all the virtual items combined

Since the table element at any given time only contains the visible rows (plus overscan) the table itself has a height smaller than the wrapper div (nr. 2). This causes the sticky header to disappear when you scroll down since the sticky header cannot escape the table element.

### Your minimal, reproducible example

https://codesandbox.io/p/devbox/tanstack-react-virtual-example-dynamic-mr8t3x

### Steps to reproduce

1. Add `position: sticky` to the `thead` element
2. Scroll down the table and watch how the header dissapears

### Expected behavior

The header should stay on top since it is sticky.

### How often does this bug happen?

Every time

### Platform

Chrome

### tanstack-virtual version

3.0.0-beta.68

### TypeScript version

5.3.2

### Additional context

I've tried getting rid of the wrapper div (nr. 2) and setting the `height: `${getTotalSize()}px` directly on the table element but this causes the rows' height the grow because there are only ever enough rows to fit on the screen (plus overscan) but having the table the full height causes the rows to evenly divide the space between each other (making them a lot larger).

### Terms & Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I understand that if my bug cannot be reliable reproduced in a debuggable environment, it will probably not be fixed and this issue may even be closed.

**Comment by ryanjoycemacq** [comment]
This happens for me too, although I put position: sticky and top: 0 in my tr, not the thead. But still able to reproduce it in your example.

**Comment by riptusk331** [comment]
I found a way to fix this, and am linking a working example. The trick is to calculate the height difference between the table element and the scrollable div, and then append an invisible pseudo-element to the table with that calculated height. This results in the table having the same height as the scrollable div, and stops the header from vanishing since you'll no longer scroll past the "end" of table element.

Doing this is a little bit tricky since pseudo elements can't be controlled directly through React/JSX inline styling, and directly adding new styles to the DOM is super expensive and forces the browser to recalculate the layout (which will make scrolling janky). The best way I found to do this was with with CSS variables, which the browser is very efficient at handling and it doesn't force a full layout recalc. I added a `virtual-table::after` CSS class with a variable height, and made sure to class my `<table>` element with `virtual-table`. Then just set that height from your handlers in your React table component.

Also, since the number of table rows (`<tr>`) being rendered to the DOM by the virtualizer at any given time varies (and therefore the height of the table + your originally set pseudo element varies), you need to pay attention to this as you near the bottom of your scrollable area as this could result in a large divergence (meaning you'll over-scroll past the table, or the pseudo element will be too short and the header will disappear again). The way I did this was set up an event listener for the scroll that flags when we're near the bottom 5% of the list, and triggers the pseudo element height recalculation.

[Sticky Table Header Example](https://codesandbox.io/p/devbox/zealous-platform-j349gw?file=%2Fsrc%2Fmain.tsx%3A22%2C26&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clqema1iz0007356gdu8fnlcv%2522%252C%2522sizes%2522%253A%255B70.23557126030624%252C29.764428739693756%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clqema1iz0002356gkmdrrs8j%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clqema1iz0004356gxjl461xf%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clqema1iz0006356g93d0yscw%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B52.084410981474036%252C47.915589018525964%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clqema1iz0002356gkmdrrs8j%2522%253A%257B%2522id%2522%253A%2522clqema1iz0002356gkmdrrs8j%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqema1iy0001356g5t9fnj7m%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252F.codesandbox%252Ftasks.json%2522%252C%2522id%2522%253A%2522clqomlxjq00fb356hzi52gyr5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clqomlxjq00fb356hzi52gyr5%2522%257D%252C%2522clqema1iz0006356g93d0yscw%2522%253A%257B%2522id%2522%253A%2522clqema1iz0006356g93d0yscw%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqema1iz0005356gxokgz940%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clqema1iz0005356gxokgz940%2522%257D%252C%2522clqema1iz0004356gxjl461xf%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqema1iz0003356g0ed7iecz%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522id%2522%253A%2522clqema1iz0004356gxjl461xf%2522%252C%2522activeTabId%2522%253A%2522clqema1iz0003356g0ed7iecz%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

I'm sure this can be improved and there are further efficiencies, but it works buttery smooth for me.

As for the "issue" itself, I would argue that this isn't really a `react-virtual` issue. I don't think there's anything the library could really do for you. This is just a quirk of virtualizing that you need to account for. Virtualizing tables is kind of a bastardization of things. That said, it would be good to include this as an example in the existing docs, as none of this is immediately obvious.

**Comment by ryanjoycemacq** [comment]
also i think the example you linked is just the regular virtual table example...there's nothing with sticky headers in it.

**Comment by wjthieme** [comment]
@ryanjoycemacq if you add `position: sticky` to the `<thead>` element in the example you can reproduce this.

@riptusk331 I'll check if this works!

**Comment by aronedwards** [comment]
Cleaner "hack/solution", but not super smooth
can use parentRef (ref of scrolling component) to generate your own code version of sticky
less performant, but more maintainable until bug solved 
```
<TableRow
    key={headerGroup.id}
    className="relative z-20"
    style={{
      transform: `translateY(${
        parentRef.current?.scrollTop || 0
      }px)`,
    }} >

**Comment by piecyk** [comment]
There are few issues here, overfall tables don't quite like when rows are positions absolute, one option is to change the positioning strategy by adding before/after row that will move items when scrolling

https://codesandbox.io/p/devbox/zealous-firefly-8vm5y2?file=%2Fsrc%2Fmain.tsx%3A91%2C19

btw there is no clear answer what positioning strategy is better, it's really hard to compare performance in a significant way.

**Comment by kelvinfloresta** [comment]
Try applying the transform only if you are not scrolling and add a smooth transition when the header appears

```
const headerPosition = virtualizer.scrollOffset

<table style="position: relative">
<thead>
<th
    key={headerGroup.id}
    style={{
      position: "relative",
      zIndex: 2,
      transition: "transform 0.5s",
      transform: !virtualizer.isScrolling
        ? `translateY(${headerPosition}px)`
        : "translateY(0px)",
    }}
  >
```

**Comment by Lipus86** [comment]
@kelvinfloresta 
if you scroll up, header is detached and shown in the middle

**Comment by bgrgv** [comment]
Here's a clean and working solution:

Separate the TableHead and the TableBody and add toggle (or other) buttons as required.

Sample implementation with Search Bar and Column Visibility toggles:

```
"use client";

import React, { useState } from "react";
import {
  ColumnDef,
  flexRender,
  getCoreRowModel,
  getFilteredRowModel,
  getPaginationRowModel,
  useReactTable,
  VisibilityState,
  SortingState,
  getSortedRowModel,
} from "@tanstack/react-table";
import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableHeader,
  TableRow,
} from "@/src/components/ui/table"; // Adjust the import path
import {
  DropdownMenu,
  DropdownMenuCheckboxItem,
  DropdownMenuContent,
  DropdownMenuTrigger,
} from "@/src/components/ui/dropdown-menu";
import { Input } from "./input"; // Adjust the import path
import { Button } from "./button"; // Adjust the import path
import { ScrollArea, ScrollBar } from "./scroll-area"; // Adjust the import path

interface DataTableProps<TData, TValue> {
  columns: ColumnDef<TData, TValue>[];
  data: TData[];
  searchKey: string;
  searchKeyLabel: string;
}

export function DataTable<TData, TValue>({
  columns,
  data,
  searchKey,
  searchKeyLabel,
}: DataTableProps<TData, TValue>) {
  const [sorting, setSorting] = useState<SortingState>([]);
  const [columnVisibility, setColumnVisibility] = useState<VisibilityState>({});

  const table = useReactTable({
    data,
    columns,
    initialState: { pagination: { pageSize: 13 } },
    getPaginationRowModel: getPaginationRowModel(),
    getCoreRowModel: getCoreRowModel(),
    getFilteredRowModel: getFilteredRowModel(),
    getSortedRowModel: getSortedRowModel(),
    onColumnVisibilityChange: setColumnVisibility,
    onSortingChange: setSorting,
    state: { sorting, columnVisibility },
  });

  return (
    <>
      <div className="flex items-center justify-between mb-4">
        <Input
          placeholder={`Search ${searchKeyLabel}...`}
          value={(table.getColumn(searchKey)?.getFilterValue() as string) ?? ""}
          onChange={(event) =>
            table.getColumn(searchKey)?.setFilterValue(event.target.value)
          }
          className="w-full md:max-w-sm"
        />
        <DropdownMenu>
          <DropdownMenuTrigger asChild>
            <Button variant="outline" className="hover:scale-105 transition-transform duration-200">
              Select Columns
            </Button>
          </DropdownMenuTrigger>
          <DropdownMenuContent align="end">
            {table.getAllColumns().filter(column => column.getCanHide()).map(column => (
              <DropdownMenuCheckboxItem
                key={column.id}
                className="capitalize"
                checked={column.getIsVisible()}
                onCheckedChange={value => column.toggleVisibility(!!value)}
              >
                {column.id}
              </DropdownMenuCheckboxItem>
            ))}
          </DropdownMenuContent>
        </DropdownMenu>
      </div>

      <ScrollArea className="rounded-md border h-[calc(80vh-220px)]">
        <Table className="table-fixed">
          <TableHeader className="sticky top-0 z-10 bg-white">
            {table.getHeaderGroups().map(headerGroup => (
              <TableRow key={headerGroup.id}>
                {headerGroup.headers.map(header => (
                  <TableHead key={header.id} className="p-1 text-left" style={{ width: header.column.columnDef.meta?.width }}>
                    {header.isPlaceholder ? null : flexRender(header.column.columnDef.header, header.getContext())}
                  </TableHead>
                ))}
              </TableRow>
            ))}
          </TableHeader>
          <TableBody>
            {table.getRowModel().rows.length ? (
              table.getRowModel().rows.map(row => (
                <TableRow key={row.id}>
                  {row.getVisibleCells().map(cell => (
                    <TableCell key={cell.id} className="p-1" style={{ width: cell.column.columnDef.meta?.width }}>
                      {flexRender(cell.column.columnDef.cell, cell.getContext())}
                    </TableCell>
                  ))}
                </TableRow>
              ))
            ) : (
              <TableRow>
                <TableCell colSpan={columns.length} className="h-24 text-center">
                  No results.
                </TableCell>
              </TableRow>
            )}
          </TableBody>
        </Table>
        <ScrollBar orientation="horizontal" />
      </ScrollArea>

      <div className="flex items-center justify-end space-x-2 py-4">
        <div className="flex-1 text-sm text-muted-foreground">
          {table.getFilteredSelectedRowModel().rows.length} of{" "}
          {table.getFilteredRowModel().rows.length} row(s) selected.
        </div>
        <Button
          variant="outline"
          size="sm"
          onClick={() => table.previousPage()}
          disabled={!table.getCanPreviousPage()}
        >
          Previous
        </Button>
        <Button
          variant="outline"
          size="sm"
          onClick={() => table.nextPage()}
          disabled={!table.getCanNextPage()}
        >
          Next
        </Button>
      </div>
    </>
  );
}
```

**Comment by Ghostblad3** [comment]
> Here's a clean and working solution:
> 
> Separate the TableHead and the TableBody and add toggle (or other) buttons as required.
> 
> Sample implementation with Search Bar and Column Visibility toggles:
> 
> ```
> "use client";
> 
> import React, { useState } from "react";
> import {
>   ColumnDef,
>   flexRender,
>   getCoreRowModel,
>   getFilteredRowModel,
>   getPaginationRowModel,
>   useReactTable,
>   VisibilityState,
>   SortingState,
>   getSortedRowModel,
> } from "@tanstack/react-table";
> import {
>   Table,
>   TableBody,
>   TableCell,
>   TableHead,
>   TableHeader,
>   TableRow,
> } from "@/src/components/ui/table"; // Adjust the import path
> import {
>   DropdownMenu,
>   DropdownMenuCheckboxItem,
>   DropdownMenuContent,
>   DropdownMenuTrigger,
> } from "@/src/components/ui/dropdown-menu";
> import { Input } from "./input"; // Adjust the import path
> import { Button } from "./button"; // Adjust the import path
> import { ScrollArea, ScrollBar } from "./scroll-area"; // Adjust the import path
> 
> interface DataTableProps<TData, TValue> {
>   columns: ColumnDef<TData, TValue>[];
>   data: TData[];
>   searchKey: string;
>   searchKeyLabel: string;
> }
> 
> export function DataTable<TData, TValue>({
>   columns,
>   data,
>   searchKey,
>   searchKeyLabel,
> }: DataTableProps<TData, TValue>) {
>   const [sorting, setSorting] = useState<SortingState>([]);
>   const [columnVisibility, setColumnVisibility] = useState<VisibilityState>({});
> 
>   const table = useReactTable({
>     data,
>     columns,
>     initialState: { pagination: { pageSize: 13 } },
>     getPaginationRowModel: getPaginationRowModel(),
>     getCoreRowModel: getCoreRowModel(),
>     getFilteredRowModel: getFilteredRowModel(),
>     getSortedRowModel: getSortedRowModel(),
>     onColumnVisibilityChange: setColumnVisibility,
>     onSortingChange: setSorting,
>     state: { sorting, columnVisibility },
>   });
> 
>   return (
>     <>
>       <div className="flex items-center justify-between mb-4">
>         <Input
>           placeholder={`Search ${searchKeyLabel}...`}
>           value={(table.getColumn(searchKey)?.getFilterValue() as string) ?? ""}
>           onChange={(event) =>
>             table.getColumn(searchKey)?.setFilterValue(event.target.value)
>           }
>           className="w-full md:max-w-sm"
>         />
>         <DropdownMenu>
>           <DropdownMenuTrigger asChild>
>             <Button variant="outline" className="hover:scale-105 transition-transform duration-200">
>               Select Columns
>             </Button>
>           </DropdownMenuTrigger>
>           <DropdownMenuContent align="end">
>             {table.getAllColumns().filter(column => column.getCanHide()).map(column => (
>               <DropdownMenuCheckboxItem
>                 key={column.id}
>                 className="capitalize"
>                 checked={column.getIsVisible()}
>                 onCheckedChange={value => column.toggleVisibility(!!value)}
>               >
>                 {column.id}
>               </DropdownMenuCheckboxItem>
>             ))}
>           </DropdownMenuContent>
>         </DropdownMenu>
>       </div>
> 
>       <ScrollArea className="rounded-md border h-[calc(80vh-220px)]">
>         <Table className="table-fixed">
>           <TableHeader className="sticky top-0 z-10 bg-white">
>             {table.getHeaderGroups().map(headerGroup => (
>               <TableRow key={headerGroup.id}>
>                 {headerGroup.headers.map(header => (
>                   <TableHead key={header.id} className="p-1 text-left" style={{ width: header.column.columnDef.meta?.width }}>
>                     {header.isPlaceholder ? null : flexRender(header.column.columnDef.header, header.getContext())}
>                   </TableHead>
>                 ))}
>               </TableRow>
>             ))}
>           </TableHeader>
>           <TableBody>
>             {table.getRowModel().rows.length ? (
>               table.getRowModel().rows.map(row => (
>                 <TableRow key={row.id}>
>                   {row.getVisibleCells().map(cell => (
>                     <TableCell key={cell.id} className="p-1" style={{ width: cell.column.columnDef.meta?.width }}>
>                       {flexRender(cell.column.columnDef.cell, cell.getContext())}
>                     </TableCell>
>                   ))}
>                 </TableRow>
>               ))
>             ) : (
>               <TableRow>
>                 <TableCell colSpan={columns.length} className="h-24 text-center">
>                   No results.
>                 </TableCell>
>               </TableRow>
>             )}
>           </TableBody>
>         </Table>
>         <ScrollBar orientation="horizontal" />
>       </ScrollArea>
> 
>       <div className="flex items-center justify-end space-x-2 py-4">
>         <div className="flex-1 text-sm text-muted-foreground">
>           {table.getFilteredSelectedRowModel().rows.length} of{" "}
>           {table.getFilteredRowModel().rows.length} row(s) selected.
>         </div>
>         <Button
>           variant="outline"
>           size="sm"
>           onClick={() => table.previousPage()}
>           disabled={!table.getCanPreviousPage()}
>         >
>           Previous
>         </Button>
>         <Button
>           variant="outline"
>           size="sm"
>           onClick={() => table.nextPage()}
>           disabled={!table.getCanNextPage()}
>         >
>           Next
>         </Button>
>       </div>
>     </>
>   );
> }
> ```

This doesn't use virtualization.

**Comment by Ryanjso** [comment]
> https://codesandbox.io/p/devbox/zealous-firefly-8vm5y2?file=%2Fsrc%2Fmain.tsx%3A91%2C19

Could you possibly add the relevant code to this thread? Your link doesn't work for me

**Comment by jiheapark** [comment]
@wjthieme 
Did you solve this problem?
I have same one with you.
I want to make table header sticky, but it does not work.

**Comment by Ghostblad3** [comment]
> > https://codesandbox.io/p/devbox/zealous-firefly-8vm5y2?file=%2Fsrc%2Fmain.tsx%3A91%2C19
> 
> Could you possibly add the relevant code to this thread? Your link doesn't work for me

```ts
import * as React from 'react'
import { createRoot } from 'react-dom/client'

import { useVirtualizer, notUndefined } from '@tanstack/react-virtual'
import {
  ColumnDef,
  flexRender,
  getCoreRowModel,
  getSortedRowModel,
  Row,
  SortingState,
  useReactTable,
} from '@tanstack/react-table'
import { makeData, Person } from './makeData'
import './index.css'

function ReactTableVirtualized() {
  const [sorting, setSorting] = React.useState<SortingState>([])

  const columns = React.useMemo<ColumnDef<Person>[]>(
    () => [
      {
        accessorKey: 'id',
        header: 'ID',
        size: 60,
      },
      {
        accessorKey: 'firstName',
        cell: (info) => info.getValue(),
      },
      {
        accessorFn: (row) => row.lastName,
        id: 'lastName',
        cell: (info) => info.getValue(),
        header: () => <span>Last Name</span>,
      },
      {
        accessorKey: 'age',
        header: () => 'Age',
        size: 50,
      },
      {
        accessorKey: 'visits',
        header: () => <span>Visits</span>,
        size: 50,
      },
      {
        accessorKey: 'status',
        header: 'Status',
      },
      {
        accessorKey: 'progress',
        header: 'Profile Progress',
        size: 80,
      },
      {
        accessorKey: 'createdAt',
        header: 'Created At',
        cell: (info) => info.getValue<Date>().toLocaleString(),
      },
    ],
    [],
  )

  const [data, setData] = React.useState(() => makeData(50_000))

  const table = useReactTable({
    data,
    columns,
    state: {
      sorting,
    },
    onSortingChange: setSorting,
    getCoreRowModel: getCoreRowModel(),
    getSortedRowModel: getSortedRowModel(),
    debugTable: true,
  })

  const { rows } = table.getRowModel()

  const parentRef = React.useRef<HTMLDivElement>(null)

  const virtualizer = useVirtualizer({
    count: rows.length,
    getScrollElement: () => parentRef.current,
    estimateSize: () => 34,
    overscan: 0,
  })
  const items = virtualizer.getVirtualItems()
  const [before, after] =
  items.length > 0
    ? [
        notUndefined(items[0]).start - virtualizer.options.scrollMargin,
        virtualizer.getTotalSize() - notUndefined(items[items.length - 1]).end,
      ]
    : [0, 0]
  const colSpan = 8

  return (
    <div ref={parentRef} className="container">
      <table>
        <thead>
          {table.getHeaderGroups().map((headerGroup) => (
            <tr key={headerGroup.id}>
              {headerGroup.headers.map((header) => {
                return (
                  <th
                    key={header.id}
                    colSpan={header.colSpan}
                    style={{ 
                      width: header.getSize(),
                      position: 'sticky',
                      top: 0,
                      backgroundColor: 'red'
                    }}
                  >
                    {header.isPlaceholder ? null : (
                      <div
                        {...{
                          className: header.column.getCanSort()
                            ? 'cursor-pointer select-none'
                            : '',
                          onClick: header.column.getToggleSortingHandler(),
                        }}
                      >
                        {flexRender(
                          header.column.columnDef.header,
                          header.getContext(),
                        )}
                        {{
                          asc: ' 🔼',
                          desc: ' 🔽',
                        }[header.column.getIsSorted() as string] ?? null}
                      </div>
                    )}
                  </th>
                )
              })}
            </tr>
          ))}
        </thead>
        <tbody>
          {before > 0 && (
            <tr>
              <td colSpan={colSpan} style={{ height: before }} />
            </tr>
          )}
          {items.map((virtualRow, index) => {
            const row = rows[virtualRow.index] as Row<Person>
            return (
              <tr
                key={row.id}
                style={{
                  height: `${virtualRow.size}px`,
                }}
              >
                {row.getVisibleCells().map((cell) => {
                  return (
                    <td key={cell.id}>
                      {flexRender(
                        cell.column.columnDef.cell,
                        cell.getContext(),
                      )}
                    </td>
                  )
                })}
              </tr>
            )
          })}
          {after > 0 && (
            <tr>
              <td colSpan={colSpan} style={{ height: after }} />
            </tr>
          )}
        </tbody>
      </table>
    </div>
  )
}

function App() {
  return (
    <div>
      <p>
        For tables, the basis for the offset of the translate css function is
        from the row's initial position itself. Because of this, we need to
        calculate the translateY pixel count different and base it off the the
        index.
      </p>
      <ReactTableVirtualized />
      <br />
      <br />
      {process.env.NODE_ENV === 'development' ? (
        <p>
          <strong>Notice:</strong> You are currently running React in
          development mode. Rendering performance will be slightly degraded
          until this application is build for production.
        </p>
      ) : null}
    </div>
  )
}

const container = document.getElementById('root')
const root = createRoot(container!)
const { StrictMode } = React

root.render(
  <StrictMode>
    <App />
  </StrictMode>,
)
```

**Comment by aman-jha-gsc** [comment]
Here is the working approach by using scrollArea from shadcn ui
`import { Card } from 'components/ui';
import { flexRender } from '@tanstack/react-table';
import { DataTablePagination } from 'components/dataTable/data-table-pagination';
import { DataTableViewOptions } from 'components/dataTable/data-table-view-options';
import { ScrollArea, ScrollBar } from 'components/ui/scroll-area';

import { Table, TableBody, TableCell, TableHead, TableHeader, TableRow } from 'components/ui/table';

const DataTableCard = ({ title, description, table, columns, toolbar, actionButtons, height = '300px', children }) => {
  return (
    <Card className="h-full flex flex-col overflow-hidden">
      <div className="bg-white">
        <div className="space-y-1.5 p-3 flex flex-row items-start bg-muted/50">
          <div className="grid gap-0.5">
            <h3 className="font-semibold tracking-tight group flex items-center gap-2 text-lg">{title}</h3>
            {description && <p className="text-sm text-muted-foreground">{description}</p>}
          </div>
        </div>
        <div className="space-x-4 p-3 flex items-center justify-between">
          {toolbar && <div className="flex items-center">{toolbar}</div>}
          <div className="ml-auto flex items-center gap-2">
            <DataTableViewOptions table={table} />
            {actionButtons}
          </div>
        </div>
      </div>
      <div className="flex-grow overflow-hidden flex flex-col">
        <div className="flex-grow">
          <Table>
            <ScrollArea className={` h-[calc(95vh-${height})]`}>
              <TableHeader className="sticky top-0 z-10 bg-[#c4eaf5]">
                {table.getHeaderGroups().map((headerGroup) => (
                  <TableRow key={headerGroup.id}>
                    {headerGroup.headers.map((header) => {
                      // Check if this header is a group
                      const isGroup = header.column.columnDef.columns !== undefined;
                      return (
                        <TableHead
                          className={` text-blue-800 font-bold p-2 text-left ${isGroup > 0 ? 'border-l-2 border-blue-300' : ''}`}
                          key={header.id}
                          colSpan={isGroup ? header.column.columnDef.columns.length : 1}
                        >
                          {header.isPlaceholder ? null : flexRender(header.column.columnDef.header, header.getContext())}
                        </TableHead>
                      );
                    })}
                  </TableRow>
                ))}
              </TableHeader>
              <TableBody>
                {table.getRowModel().rows?.length ? (
                  table.getRowModel().rows.map((row) => (
                    <TableRow className="even:bg-slate-100" key={row.id} data-state={row.getIsSelected() && 'selected'}>
                      {row.getVisibleCells().map((cell) => (
                        <TableCell key={cell.id} className="p-1" style={{ width: cell.column.columnDef.meta?.width }}>
                          {flexRender(cell.column.columnDef.cell, cell.getContext())}
                        </TableCell>
                      ))}
                    </TableRow>
                  ))
                ) : (
                  <TableRow>
                    <TableCell colSpan={columns.length} className="h-24 text-center">
                      No results.
                    </TableCell>
                  </TableRow>
                )}
              </TableBody>
              <ScrollBar orientation="horizontal" />
            </ScrollArea>
          </Table>
        </div>
      </div>
      <div className="mt-auto p-3">
        <DataTablePagination table={table} />
      </div>
    </Card>
  );
};

export default DataTableCard;
`

**Comment by aman-jha-gsc** [comment]
Here is the working code for your reference

```
import { Card } from '@metro-ui/core';
import { flexRender } from '@tanstack/react-table';
import { DataTablePagination } from 'components/dataTable/data-table-pagination';
import { DataTableViewOptions } from 'components/dataTable/data-table-view-options';
import { ScrollArea, ScrollBar } from 'components/ui/scroll-area';
import { Table, TableBody, TableCell, TableHead, TableHeader, TableRow } from 'components/ui/table';

const DataTableCard = ({ title, description, table, columns, toolbar, actionButtons, height = '300px', children }) => {
  return (
    <Card className="h-full flex flex-col overflow-hidden">
      <div className="bg-white">
        <div className="space-y-1.5 p-3 flex flex-row items-start bg-muted/50">
          <div className="grid gap-0.5">
            <h3 className="font-semibold tracking-tight group flex items-center gap-2 text-lg">{title}</h3>
            {description && <p className="text-sm text-muted-foreground">{description}</p>}
          </div>
        </div>
        <div className="space-x-4 p-3 flex items-center justify-between">
          {toolbar && <div className="flex items-center">{toolbar}</div>}
          <div className="ml-auto flex items-center gap-2">
            <DataTableViewOptions table={table} />
            {actionButtons}
          </div>
        </div>
      </div>
      <div className="flex-grow overflow-hidden flex flex-col">
        <div className="flex-grow">
          <Table>
            <ScrollArea className={` h-[calc(95vh-${height})]`}>
              <TableHeader className="sticky top-0 z-10 bg-[#c4eaf5]">
                {table.getHeaderGroups().map((headerGroup) => (
                  <TableRow key={headerGroup.id}>
                    {headerGroup.headers.map((header) => {
                      // Check if this header is a group
                      const isGroup = header.column.columnDef.columns !== undefined;
                      return (
                        <TableHead
                          className={` text-blue-800 font-bold p-2 text-left ${isGroup > 0 ? 'border-l-2 border-blue-300' : ''}`}
                          key={header.id}
                          colSpan={isGroup ? header.column.columnDef.columns.length : 1}
                        >
                          {header.isPlaceholder ? null : flexRender(header.column.columnDef.header, header.getContext())}
                        </TableHead>
                      );
                    })}
                  </TableRow>
                ))}
              </TableHeader>
              <TableBody>
                {table.getRowModel().rows?.length ? (
                  table.getRowModel().rows.map((row) => (
                    <TableRow className="even:bg-slate-100" key={row.id} data-state={row.getIsSelected() && 'selected'}>
                      {row.getVisibleCells().map((cell) => (
                        <TableCell key={cell.id} className="p-1" style={{ width: cell.column.columnDef.meta?.width }}>
                          {flexRender(cell.column.columnDef.cell, cell.getContext())}
                        </TableCell>
                      ))}
                    </TableRow>
                  ))
                ) : (
                  <TableRow>
                    <TableCell colSpan={columns.length} className="h-24 text-center">
                      No results.
                    </TableCell>
                  </TableRow>
                )}
              </TableBody>
              <ScrollBar orientation="horizontal" />
            </ScrollArea>
          </Table>
        </div>
      </div>
      <div className="mt-auto p-3">
        <DataTablePagination table={table} />
      </div>
    </Card>
  );
};

export default DataTableCard;

```

**Comment by damiaomk** [comment]
> ### Describe the bug
> When using a react-table together with vritualizer and a sticky header the sticky header disappears when scrolling. This is due to the fact that when combining the two the table should be wrapped in two divs:
> 
> 1. The container that takes the height that should be taken up. This is the div that is scrollable
> 2. The div that directly wraps the container and takes the height of all the virtual items combined
> 
> Since the table element at any given time only contains the visible rows (plus overscan) the table itself has a height smaller than the wrapper div (nr. 2). This causes the sticky header to disappear when you scroll down since the sticky header cannot escape the table element.
> 
> ### Your minimal, reproducible example
> https://codesandbox.io/p/devbox/tanstack-react-virtual-example-dynamic-mr8t3x
> 
> ### Steps to reproduce
> 1. Add `position: sticky` to the `thead` element
> 2. Scroll down the table and watch how the header dissapears
> 
> ### Expected behavior
> The header should stay on top since it is sticky.
> 
> ### How often does this bug happen?
> Every time
> 
> ### Screenshots or Videos
> _No response_
> 
> ### Platform
> Chrome
> 
> ### tanstack-virtual version
> 3.0.0-beta.68
> 
> ### TypeScript version
> 5.3.2
> 
> ### Additional context
> I've tried getting rid of the wrapper div (nr. 2) and setting the `height: `${getTotalSize()}px` directly on the table element but this causes the rows' height the grow because there are only ever enough rows to fit on the screen (plus overscan) but having the table the full height causes the rows to evenly divide the space between each other (making them a lot larger).
> 
> ### Terms & Code of Conduct
> * [x]  I agree to follow this project's Code of Conduct[x]  I understand that if my bug cannot be reliable reproduced in a debuggable environment, it will probably not be fixed and this issue may even be closed.

Did you solve this problem ?

**Comment by newbeelearn** [comment]
here is the snippet that works, i have just copy, pasted did not cleaned it up so please ignore the irrelevant bits.
Key bits are 

style={{
		height: `${virtualizer.getTotalSize()}px`
              }}> 

on tbody this will ensure that your scrollbar indicates true size of table and if you use sticky in header it will not go away.

<tr
                    key={virtualRow.index}
                    className="border-b transition-colors hover:bg-muted/50 h-12 absolute w-full"
                    style={{
                      transform: `translateY(${virtualRow.start}px)`,
                    }}
                  >
This will ensure that your virtualized rows are correctly positioned.

Last but not the least after this if you have sufficiently large table(million+ rows) you will hit browsers css pixel limit, you can refer to this issue https://github.com/TanStack/virtual/issues/460 . If its not the case than the current solution should work.

with the way current lib is setup, i am not sure 

(
        <div ref={parentRef} className="h-[600px] overflow-auto relative">
          <table className="w-full caption-bottom text-sm border">
            <thead className="[&_tr]:border-b">
              <tr className="border-b transition-colors hover:bg-muted/50">
                {visibleColumns
                  .map((column) => (
                    <th
                      key={column.id}
		      ref={(el) => (headerRefs.current[column.id] = el)}
                      className="h-12 px-4 text-left font-medium text-muted-foreground sticky z-10 top-0 bg-background"
                    >
                      <HeaderCell
                        column={column}
                        onSort={() => handleSort(column.id)}
                        sortDirection={ sorting.find(s => s.column === column.id)?.operator || 'none' }
                      />
                    </th>
                  ))}
              </tr>
            </thead>
            <tbody className={tableMetadata.styles?.tableBody ?? `p-4 relative w-full`}
	    style={{
		height: `${virtualizer.getTotalSize()}px`
              }}>
              {virtualizer.getVirtualItems().map((virtualRow) => {
                const row = tableData[virtualRow.index];
                return (
                  <tr
                    key={virtualRow.index}
                    className="border-b transition-colors hover:bg-muted/50 h-12 absolute w-full"
                    style={{
                      transform: `translateY(${virtualRow.start}px)`,
                    }}
                  >
                    {visibleColumns.map((column) => (
                      <td key={column.id} className="px-4 py-1 text-left"
		      style={{ width: `${columnWidths[column.id]}px` }}>
                        <p className="max-w-full h-10 overflow-auto overflow-y-auto whitespace-normal"
                          title={String(getValue(dataTypes, column.id, row))}>
                          {String(getValue(dataTypes, column.id, row))}
                        </p>
                      </td>
                    ))}
                  </tr>
                );
              })}
            </tbody>
          </table>
        </div>
	)

**Comment by piecyk** [comment]
@newbeelearn Thanks for sharing the snippet! You're right, the current implementation doesn't handle the maximum pixel limit, and this is something that can happen with very large tables. Handling this is not trivial, as it introduces additional complexity with managing large offsets.

That said, we’re open to ideas and would be happy to accept any PR proposals addressing this. If you have a solution in mind, feel free to contribute!

**Comment by seedydocloop** [comment]
@newbeelearn Thanks for sharing your solution !

Sadly, for tables with fixed-layout, setting `position: 'absolute'` breaks column layout.

Is there any solution with relatively positioned table rows ?

**Comment by MKSisti** [comment]
> [@newbeelearn](https://github.com/newbeelearn) Thanks for sharing your solution !
> 
> Sadly, for tables with fixed-layout, setting `position: 'absolute'` breaks column layout.
> 
> Is there any solution with relatively positioned table rows ?

Did you find a solution for this ? @seedydocloop, been stuck on this for a while now

**Comment by piecyk** [comment]
@seedydocloop @MKSisti relatively positioned rows directly usually won’t work as intended because tables do not behave like regular block elements.

**Comment by seedydocloop** [comment]
@piecyk @MKSisti I ended up using a before and after row which height simulate the missing space.
This way, rows are 'position: static'.
Found this in another discussion.

**Comment by MKSisti** [comment]
Thank you for replying 🙏 
@piecyk, makes sense but what would be a good approach to having virtualized tables that keep the column layout if there's any .
@seedydocloop as in injecting a before row in the data ? or adding a row before the virtual rows?, can you share a link to the discussion might be helpful, thanks a lot in advance.

**Comment by soliyapintu** [comment]
@piecyk Do you have any suggestions or solutions for implementing sticky columns when using horizontal virtualization in @tanstack/react-table?

**Comment by houfeng0923** [comment]
we can add prepend and append empty `<tr>` in `<tbody>` with  computed height.
or add extra <tbody empty></tbody><tbody data><tbody><tbody empty></tbody>

pls check this demo:

https://stackblitz.com/edit/tanstack-virtual-wrmyygfq?file=src%2FApp.vue

and 

https://stackblitz.com/edit/tanstack-virtual-nbspetuk?file=src%2FApp.vue

**Comment by hackel** [comment]
Using the implementation described here by @piecyk and @houfeng0923, there is a bug when the number of header rows exceeds the scroll element height.  You can see this demonstrated here:

https://stackblitz.com/edit/tanstack-virtual-nphdztkk?file=src%2FApp.vue,src%2Fstyle.css

When scrolling down past the header rows, the "before" TR is already so large that you only see white space instead of the first data rows.  Only after scrolling all the way to the bottom can you see any actual data.

Any insights into how to solve this problem?

**Comment by houfeng0923** [comment]
> Using the implementation described here by [@piecyk](https://github.com/piecyk) and [@houfeng0923](https://github.com/houfeng0923), there is a bug when the number of header rows exceeds the scroll element height. You can see this demonstrated here:
> 
> https://stackblitz.com/edit/tanstack-virtual-nphdztkk?file=src%2FApp.vue,src%2Fstyle.css
> 
> When scrolling down past the header rows, the "before" TR is already so large that you only see white space instead of the first data rows. Only after scrolling all the way to the bottom can you see any actual data.
> 
> Any insights into how to solve this problem?
looks header cell wrap , try: 
th {white-space: nowrap;}

https://stackblitz.com/edit/tanstack-virtual-djthtfan?file=src%2FApp.vue,src%2Fstyle.css

**Comment by hackel** [comment]
@houfeng0923 Thanks, but even in your example, once you scroll down a few rows you start to see a changing 1-2 line gap above the tbody.

I noticed that if you crank up the overscan to 25, the problem goes away, but this slows down the rendering quite a bit.

https://stackblitz.com/edit/tanstack-virtual-gce1xdfk?file=src%2FApp.vue,src%2Fstyle.css

**Comment by Himan8077** [comment]
@riptusk331 i tried what you mentioned and it helped me with the sticky header thing, but now i'm facing another issue and it is browser specific(facing issue on safari only) as when i'm at near bottom scroll bar sometime moved back to the top of the table or it start fickering and last row is not scrollable, if you have any idea please let me know about this

**Comment by idan-h** [comment]
> [@riptusk331](https://github.com/riptusk331) i tried what you mentioned and it helped me with the sticky header thing, but now i'm facing another issue and it is browser specific(facing issue on safari only) as when i'm at near bottom scroll bar sometime moved back to the top of the table or it start fickering and last row is not scrollable, if you have any idea please let me know about this

This might not be related to this problem, I had the same thing when I set the     `estimateSize: () => ...,` to a smaller size than the actual row height.

**Comment by idan-h** [comment]
> <table className="w-full caption-bottom text-sm border">

this does not work at all. what I got was scary big rows that gets bigger as you scroll, and also the thead still disapears haha

**Comment by idan-h** [comment]
> here is the snippet that works, i have just copy, pasted did not cleaned it up so please ignore the irrelevant bits. Key bits are
> 
> ```
> style={{
> 		height: `${virtualizer.getTotalSize()}px`
>               }}> 
> ```
> 
> on tbody this will ensure that your scrollbar indicates true size of table and if you use sticky in header it will not go away.
> 
> ```
> <tr
>                     key={virtualRow.index}
>                     className="border-b transition-colors hover:bg-muted/50 h-12 absolute w-full"
>                     style={{
>                       transform: `translateY(${virtualRow.start}px)`,
>                     }}
>                   >
> ```
> 
> This will ensure that your virtualized rows are correctly positioned.
> 
> Last but not the least after this if you have sufficiently large table(million+ rows) you will hit browsers css pixel limit, you can refer to this issue [#460](https://github.com/TanStack/virtual/issues/460) . If its not the case than the current solution should work.
> 
> with the way current lib is setup, i am not sure if there is a solution with pixel based approach
> 
> ```
> (
>         <div ref={parentRef} className="h-[600px] overflow-auto relative">
>           <table className="w-full caption-bottom text-sm border">
>             <thead className="[&_tr]:border-b">
>               <tr className="border-b transition-colors hover:bg-muted/50">
>                 {visibleColumns
>                   .map((column) => (
>                     <th
>                       key={column.id}
> 		      ref={(el) => (headerRefs.current[column.id] = el)}
>                       className="h-12 px-4 text-left font-medium text-muted-foreground sticky z-10 top-0 bg-background"
>                     >
>                       <HeaderCell
>                         column={column}
>                         onSort={() => handleSort(column.id)}
>                         sortDirection={ sorting.find(s => s.column === column.id)?.operator || 'none' }
>                       />
>                     </th>
>                   ))}
>               </tr>
>             </thead>
>             <tbody className={tableMetadata.styles?.tableBody ?? `p-4 relative w-full`}
> 	    style={{
> 		height: `${virtualizer.getTotalSize()}px`
>               }}>
>               {virtualizer.getVirtualItems().map((virtualRow) => {
>                 const row = tableData[virtualRow.index];
>                 return (
>                   <tr
>                     key={virtualRow.index}
>                     className="border-b transition-colors hover:bg-muted/50 h-12 absolute w-full"
>                     style={{
>                       transform: `translateY(${virtualRow.start}px)`,
>                     }}
>                   >
>                     {visibleColumns.map((column) => (
>                       <td key={column.id} className="px-4 py-1 text-left"
> 		      style={{ width: `${columnWidths[column.id]}px` }}>
>                         <p className="max-w-full h-10 overflow-auto overflow-y-auto whitespace-normal"
>                           title={String(getValue(dataTypes, column.id, row))}>
>                           {String(getValue(dataTypes, column.id, row))}
>                         </p>
>                       </td>
>                     ))}
>                   </tr>
>                 );
>               })}
>             </tbody>
>           </table>
>         </div>
> 	) 
> ```

this does not work at all. what I got was scary big rows that gets bigger as you scroll, and also the thead still disapears haha

**Comment by Himan8077** [comment]
@idan-h although the issue i mentioned is only happening in Safari, in other browser it is working as expected

**Comment by RobertWadeSteele** [comment]
> I found a way to fix this, and am linking a working example. The trick is to calculate the height difference between the table element and the scrollable div, and then append an invisible pseudo-element to the table with that calculated height. This results in the table having the same height as the scrollable div, and stops the header from vanishing since you'll no longer scroll past the "end" of table element.
> 
> Doing this is a little bit tricky since pseudo elements can't be controlled directly through React/JSX inline styling, and directly adding new styles to the DOM is super expensive and forces the browser to recalculate the layout (which will make scrolling janky). The best way I found to do this was with with CSS variables, which the browser is very efficient at handling and it doesn't force a full layout recalc. I added a `virtual-table::after` CSS class with a variable height, and made sure to class my `<table>` element with `virtual-table`. Then just set that height from your handlers in your React table component.
> 
> Also, since the number of table rows (`<tr>`) being rendered to the DOM by the virtualizer at any given time varies (and therefore the height of the table + your originally set pseudo element varies), you need to pay attention to this as you near the bottom of your scrollable area as this could result in a large divergence (meaning you'll over-scroll past the table, or the pseudo element will be too short and the header will disappear again). The way I did this was set up an event listener for the scroll that flags when we're near the bottom 5% of the list, and triggers the pseudo element height recalculation.
> 
> [Sticky Table Header Example](https://codesandbox.io/p/devbox/zealous-platform-j349gw?file=%2Fsrc%2Fmain.tsx%3A22%2C26&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clqema1iz0007356gdu8fnlcv%2522%252C%2522sizes%2522%253A%255B70.23557126030624%252C29.764428739693756%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clqema1iz0002356gkmdrrs8j%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clqema1iz0004356gxjl461xf%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clqema1iz0006356g93d0yscw%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B52.084410981474036%252C47.915589018525964%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clqema1iz0002356gkmdrrs8j%2522%253A%257B%2522id%2522%253A%2522clqema1iz0002356gkmdrrs8j%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqema1iy0001356g5t9fnj7m%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252F.codesandbox%252Ftasks.json%2522%252C%2522id%2522%253A%2522clqomlxjq00fb356hzi52gyr5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clqomlxjq00fb356hzi52gyr5%2522%257D%252C%2522clqema1iz0006356g93d0yscw%2522%253A%257B%2522id%2522%253A%2522clqema1iz0006356g93d0yscw%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqema1iz0005356gxokgz940%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clqema1iz0005356gxokgz940%2522%257D%252C%2522clqema1iz0004356gxjl461xf%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqema1iz0003356g0ed7iecz%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522id%2522%253A%2522clqema1iz0004356gxjl461xf%2522%252C%2522activeTabId%2522%253A%2522clqema1iz0003356g0ed7iecz%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
> 
> I'm sure this can be improved and there are further efficiencies, but it works buttery smooth for me.
> 
> As for the "issue" itself, I would argue that this isn't really a `react-virtual` issue. I don't think there's anything the library could really do for you. This is just a quirk of virtualizing that you need to account for. Virtualizing tables is kind of a bastardization of things. That said, it would be good to include this as an example in the existing docs, as none of this is immediately obvious.

This saved me @riptusk331. Thank you so much. For anyone using tailwind, here's an easy way to apply the style. I'm not sure if a better solution was proposed. I didn't really read too far after riptusks comment.

```typescript
function Example() {
    ...

    const parentRef = useRef<HTMLDivElement>(null);
    const virtualizer = useVirtualizer({
        count: rows.length,
        estimateSize: () => 59,
        getScrollElement: () => parentRef.current,
    });

    return (
        <div ref={parentRef} className="relative overflow-auto">
            <div
                style={{
                    height: `${virtualizer.getTotalSize()}px`,
                }}
            >
                <table className="after:inline-block after:h-(--my-variable)">
                    <thead className="sticky top-0 z-1">...</thead>
                    <tbody
                        ref={(ref) => {
                            if (!ref) return;

                            const height =
                                virtualizer.getTotalSize() -
                                ref.getBoundingClientRect().height;

                            document.documentElement.style.setProperty(
                                '--my-variable',
                                `${height}px`,
                            );
                        }}
                    >...</tbody>
                </table>
            </div>
        </div>
    );
}
```

**Comment by MoSattler** [comment]
I am still struggling with this. Has anyone managed to get it to work with the `useWindowVirtualizer`?

**Comment by MoSattler** [comment]
This is not only an issue with headings, but also all tbody styling.

**Comment by beckerrobin** [comment]
> > I found a way to fix this, and am linking a working example. The trick is to calculate the height difference between the table element and the scrollable div, and then append an invisible pseudo-element to the table with that calculated height. This results in the table having the same height as the scrollable div, and stops the header from vanishing since you'll no longer scroll past the "end" of table element.
> > Doing this is a little bit tricky since pseudo elements can't be controlled directly through React/JSX inline styling, and directly adding new styles to the DOM is super expensive and forces the browser to recalculate the layout (which will make scrolling janky). The best way I found to do this was with with CSS variables, which the browser is very efficient at handling and it doesn't force a full layout recalc. I added a `virtual-table::after` CSS class with a variable height, and made sure to class my `<table>` element with `virtual-table`. Then just set that height from your handlers in your React table component.
> > Also, since the number of table rows (`<tr>`) being rendered to the DOM by the virtualizer at any given time varies (and therefore the height of the table + your originally set pseudo element varies), you need to pay attention to this as you near the bottom of your scrollable area as this could result in a large divergence (meaning you'll over-scroll past the table, or the pseudo element will be too short and the header will disappear again). The way I did this was set up an event listener for the scroll that flags when we're near the bottom 5% of the list, and triggers the pseudo element height recalculation.
> > [Sticky Table Header Example](https://codesandbox.io/p/devbox/zealous-platform-j349gw?file=%2Fsrc%2Fmain.tsx%3A22%2C26&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clqema1iz0007356gdu8fnlcv%2522%252C%2522sizes%2522%253A%255B70.23557126030624%252C29.764428739693756%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clqema1iz0002356gkmdrrs8j%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clqema1iz0004356gxjl461xf%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clqema1iz0006356g93d0yscw%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B52.084410981474036%252C47.915589018525964%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clqema1iz0002356gkmdrrs8j%2522%253A%257B%2522id%2522%253A%2522clqema1iz0002356gkmdrrs8j%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqema1iy0001356g5t9fnj7m%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252F.codesandbox%252Ftasks.json%2522%252C%2522id%2522%253A%2522clqomlxjq00fb356hzi52gyr5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clqomlxjq00fb356hzi52gyr5%2522%257D%252C%2522clqema1iz0006356g93d0yscw%2522%253A%257B%2522id%2522%253A%2522clqema1iz0006356g93d0yscw%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqema1iz0005356gxokgz940%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clqema1iz0005356gxokgz940%2522%257D%252C%2522clqema1iz0004356gxjl461xf%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqema1iz0003356g0ed7iecz%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522id%2522%253A%2522clqema1iz0004356gxjl461xf%2522%252C%2522activeTabId%2522%253A%2522clqema1iz0003356g0ed7iecz%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
> > I'm sure this can be improved and there are further efficiencies, but it works buttery smooth for me.
> > As for the "issue" itself, I would argue that this isn't really a `react-virtual` issue. I don't think there's anything the library could really do for you. This is just a quirk of virtualizing that you need to account for. Virtualizing tables is kind of a bastardization of things. That said, it would be good to include this as an example in the existing docs, as none of this is immediately obvious.
> 
> This saved me [@riptusk331](https://github.com/riptusk331). Thank you so much. For anyone using tailwind, here's an easy way to apply the style. I'm not sure if a better solution was proposed. I didn't really read too far after riptusks comment.
> 
> function Example() {
>     ...
> 
>     const parentRef = useRef<HTMLDivElement>(null);
>     const virtualizer = useVirtualizer({
>         count: rows.length,
>         estimateSize: () => 59,
>         getScrollElement: () => parentRef.current,
>     });
> 
>     return (
>         <div ref={parentRef} className="relative overflow-auto">
>             <div
>                 style={{
>                     height: `${virtualizer.getTotalSize()}px`,
>                 }}
>             >
>                 <table className="after:inline-block after:h-(--my-variable)">
>                     <thead className="sticky top-0 z-1">...</thead>
>                     <tbody
>                         ref={(ref) => {
>                             if (!ref) return;
> 
>                             const height =
>                                 virtualizer.getTotalSize() -
>                                 ref.getBoundingClientRect().height;
> 
>                             document.documentElement.style.setProperty(
>                                 '--my-variable',
>                                 `${height}px`,
>                             );
>                         }}
>                     >...</tbody>
>                 </table>
>             </div>
>         </div>
>     );
> }

Vanilla CSS:
```
.your-table::after {
    height: var(--my-variable);
    display: inline-block;
    content: '';
}
```
